test/helpers/foundry/helpers.mts

```diff
@@ -0,0 +1,393 @@
+ ╠╣  ║ ║ ║ ║ ║║║  ║║ ╠╦╝ ╚╦╝    for Ethereum Application Development
+ ╚   ╚═╝ ╚═╝ ╝╚╝ ═╩╝ ╩╚═  ╩                 written in Rust.
+
+.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx
+
+Repo       : https://github.com/foundry-rs/
+Book       : https://book.getfoundry.sh/
+Chat       : https://t.me/foundry_rs/
+Support    : https://t.me/foundry_support/
+Contribute : https://github.com/orgs/foundry-rs/projects/2/
+
+.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx
+`);
+}
+
+/**
+ * Log a message to the console.
+ *
+ * @param message The message to log
+ */
+export function say(message: string) {
+  console.log(`[foundryup] ${message}`);
+}
+
+/**
+ * Extracts the binaries from the given URL and writes them to the destination.
+ *
+ * @param url The URL of the archive to extract the binaries from
+ * @param binaries The list of binaries to extract
+ * @param dir The destination directory
+ * @returns The list of binaries extracted
+ */
+export async function extractFrom(url: URL, binaries: string[], dir: string) {
+  const extract = url.pathname.toLowerCase().endsWith(BinFormat.Tar)
+    ? extractFromTar
+    : extractFromZip;
+  // write all files to a temporary directory first, then rename to the final
+  // destination to avoid accidental partial extraction. We don't use
+  // `os.tmpdir` for this because `rename` will fail if the directories are on
+  // different file systems.
+  const tempDir = dir + '.downloading';
+  const rmOpts = { recursive: true, maxRetries: 3, force: true };
+  try {
+    // clean up any previous in-progress downloads
+    await rm(tempDir, rmOpts);
+    // make the temporary directory to extract the binaries to
+    await mkdir(tempDir, { recursive: true });
+    const downloads = await extract(url, binaries, tempDir);
+    ok(downloads.length === binaries.length, 'Failed to extract all binaries');
+
+    // everything has been extracted; move the files to their final destination
+    await rename(tempDir, dir);
+    // return the list of extracted binaries
+    return downloads.map((file) => join(dir, relative(tempDir, file)));
+  } catch (e) {
+    // if things fail for any reason try to clean up a bit. it is very important
+    // to not leave `dir` behind, as its existence is a signal that the binaries
+    // are installed.
+    await Promise.all([rm(tempDir, rmOpts), rm(dir, rmOpts)]).catch(noop);
+    throw e;
+  }
+}
+
+/**
+ * Extracts the binaries from a tar archive.
+ *
+ * @param url The URL of the archive to extract the binaries from
+ * @param binaries The list of binaries to extract
+ * @param dir The destination directory
+ * @returns The list of binaries extracted
+ */
+async function extractFromTar(url: URL, binaries: string[], dir: string) {
+  const downloads: string[] = [];
+  await pipeline(
+    startDownload(url),
+    extractTar({ cwd: dir }, binaries).on('entry', ({ absolute }) =>
```

Comment:
We can use `extractTar` to only extract the binaries we asked for. So we do. Zoom zoom.